File: 그래프_dfs/leetcode472(Hard)-Concatenated Words.js
// 160ms(98.11%), 53MB(94.34%)
/**
 * @param {string[]} words
 * @return {string[]}
 */
var findAllConcatenatedWordsInADict = function (words) {
  const table = new Set(words)

  const dfs = (word, count) => {
    if (count >= 2 && word.length === 0) {
      return true
    }
    for (let i = 1; i < word.length + 1; i++) {
      const substr = word.substring(0, i)
      if (table.has(substr)) {
        if (dfs(word.slice(i), count + 1)) {
          return true
        }
      }
    }
    return false
  }

  return words.filter((word) => dfs(word, 0))
}

File: 그래프_dfs/leetcode472(Hard)-Concatenated Words.py
# 548ms(57.23%), 27.7 MB(47.98%)
class Solution:
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        table = set(words)

        def dfs(word, count):
            if count >= 2 and len(word) == 0:
                return True
            for i in range(1, len(word) + 1):
                substr = word[0 : i]
                if substr in table:
                    if dfs(word[i::], count + 1):
                        return True
            return False

        return [word for word in words if dfs(word, 0)]